public/app/plugins/panel/table/transformers.ts

```diff
@@ -135,19 +135,125 @@ transformers['table'] = {
     if (!data || data.length === 0) {
       return [];
     }
-    return data[0].columns;
+
+    // Single query returns data columns as is
+    if (data.length === 1) {
+      return [...data[0].columns];
+    }
+
+    // Track column indexes: name -> index
+    const columnNames = {};
+
+    // Union of all columns
+    const columns = data.reduce((acc, d, i) => {
+      d.columns.forEach((col, j) => {
+        const { text } = col;
+        if (columnNames[text] === undefined) {
+          columnNames[text] = acc.length;
+          acc.push(col);
+        }
+      });
+      return acc;
+    }, []);
+
+    return columns;
   },
   transform: function(data, panel, model) {
     if (!data || data.length === 0) {
       return;
     }
 
-    if (data[0].type !== 'table') {
-      throw {message: 'Query result is not in table format, try using another transform.'};
+    const noTableIndex = _.findIndex(data, d => d.type !== 'table');
+    if (noTableIndex > -1) {
+      throw {message: `Result of query #${String.fromCharCode(65 + noTableIndex)} is not in table format, try using another transform.`};
     }
 
-    model.columns = data[0].columns;
-    model.rows = data[0].rows;
+    // Single query returns data columns and rows as is
+    if (data.length === 1) {
+      model.columns = [...data[0].columns];
+      model.rows = [...data[0].rows];
+      return;
+    }
+
+    // Track column indexes: name -> index
+    const columnNames = {};
+    const columnIndexes = [];
+
+    // Union of all non-value columns
+    const columns = data.reduce((acc, d, i) => {
+      const indexes = [];
+      d.columns.forEach((col, j) => {
+        const { text } = col;
+        if (columnNames[text] === undefined) {
+          columnNames[text] = acc.length;
+          acc.push(col);
+        }
+        indexes[j] = columnNames[text];
+      });
+      columnIndexes.push(indexes);
+      return acc;
+    }, []);
+
+    model.columns = columns;
+
+    // Adjust rows to new column indexes
+    let rows = data.reduce((acc, d, i) => {
+      const indexes = columnIndexes[i];
+      d.rows.forEach((r, j) => {
+        const alteredRow = [];
+        indexes.forEach((to, from) => {
+          alteredRow[to] = r[from];
+        });
+        acc.push(alteredRow);
+      });
+      return acc;
+    }, []);
+
+    // Merge rows that have same values for columns
+    const mergedRows = {};
+    rows = rows.reduce((acc, row, rowIndex) => {
```

Comment:
I would love if this could be split up a bit more. It feels like I'm staring down the abyss :)

Reply:
Cheers! Will split this up a bit more and make it less nested.